teps/00XX-interceptor-plugins.md

```diff
@@ -0,0 +1,389 @@
+that would count as tricky in the implementation and anything particularly
+challenging to test should be called out.
+
+All code is expected to have adequate tests (eventually with coverage
+expectations).
+-->
+* Unit tests for the interceptor code
+* E2E tests for each interceptor
+
+## Drawbacks
+
+* TODO: More moving pieces -> extra deployments
+* Adding new Interceptor -> new CRD to install
+
+## Alternatives
+
+*   Use a Go plugin library for built in interceptors 
+
+    We'd leverage a library like HashiCorp's go-plugin. In this model, the interceptor will run as separate processes but within the same pod. They will communicate with the listener over RPC. So, in many ways this is similar to the proposal but the key difference is that everything runs in the same pod. This is a reasonable alternative. But it does lead us to having to maintain two different interceptor models - one for plugin one for webhooks.   
+
+*   We can keep using the same service account model as before
+*   No additional deployments
+*   Interceptor authors will have to understand the plugin library to add a new interceptor
+*   We'll still have to support both the existing Webhook interceptor model as well as the plugin based model.
+*   We'd have to figure out the contract for adding new interceptors via CRD.
+    *   one idea. the Interceptor CRD defines a containerSpec which we mount at a specific location in the EL pod.
+
+*   Keep the current model but allow operators to disable some built-in interceptors
+
+    We'll package in built-in interceptors as we do now, but an operator can turn it off using a flag or a config map. This is similar to Kubernetes where some admission controllers are [compiled in](https://kubernetes.io/docs/reference/access-authn-authz/admission-controllers/) and can only be [enabled ](https://kubernetes.io/docs/reference/access-authn-authz/admission-controllers/#how-do-i-turn-on-an-admission-controller)or disabled. Uses can extend by writing and registering their own [dynamic admission webhooks](https://kubernetes.io/docs/reference/access-authn-authz/extensible-admission-controllers/).
+
+*   Extend the existing Webhook Interface 
+
+    In this model, the built-in interceptors will still be implemented as separate services but they will use the existing Webhook Interface. Extra params from Interceptors will be sent as additional headers instead of fields in the body.
+
+
+    The main disadvantage of this is that it will be harder to add supporting fields to the response e.g. debug messages, whether the processing failed or not etc. 
+
+
+## Upgrade & Migration Strategy
+
+<!--
+Use this section to detail wether this feature needs an upgrade or
+migration strategy. This is especially useful when we modify a
+behavior or add a feature that may replace and deprecate a current one.
+-->
+
+* Old interface supported till beta
+* We can reserve the `webhook` type for callouts to webhooks
```

Comment:
can you explain a bit more about what the webhook type is?

Reply:
the current webhook interface. I changed this section to make this clearer.